Source/ParticleInterpolator.H

```diff
@@ -0,0 +1,133 @@
+#ifndef PARTICLE_INTERPOLATOR_H_
+#define PARTICLE_INTERPOLATOR_H_
+
+/*
+   The ParticleInterpolator provides cell indexing and shape
+   functions for Particle-in-Cell spline interpolation between
+   particles and an underlying regularly spaced grid.
+
+   Template specializations are provided for spline orders 0, 1, and 2.
+
+   Cell indexing is only defined for cell centered grid data,
+   and it provides cell indices relative to the leftmost cell in the domain.
+   That is, if the leftmost interior cell is index N, you will need
+   to offset the indices returned by this interpolator by +N.
+
+   This interpolator works with coordinates in units of the grid cell spacing
+   using a variable delta defined as follows:
+       delta = (x_particle - xlo)/dx
+   where:
+       * x_particle: the particle coordinate
+       * xlo: the coordinate of the left (lo) edge of the domain
+       * dx: the cell spacing
+
+   The API provided by ParticleInterpolator is:
+       * first(): returns the first cell index supporting the particle shape
+       * last(): returns the last cell index supporting the particle shape
+       * operator(int i): returns the shape function for cell index `first() <= i <= last()`
+*/
+
+#include <AMReX_REAL.H>
+
+using namespace amrex;
+
+template <int max_spline_order>
+struct ParticleInterpolator
+{
+    AMREX_GPU_DEVICE AMREX_FORCE_INLINE
+    ParticleInterpolator(Real delta, int order) {
+        // although we've set the max order at compile time we can
+        // always choose to run with a lower order at runtime.
+        if (max_spline_order >= 2 && order > 1)
+            compute_order_2(delta);
+        else if (max_spline_order >= 1 && order > 0)
+            compute_order_1(delta);
+        else
```

Comment:
My instinct here would be to use == comparisons and have the default case be an error to prevent quietly using an unexpected interpolation order. I think this is fine as is, just mentioning it in case you think it would be a good idea.

Reply:
Ah, yes, it's a great idea to check for that quiet case, would a debug mode assert be good enough?

The reason for checking based on the template argument is if, eg, you compile with shape factor 1, the compiler should be able to eliminate the first check for order 2. That gets rid of an if and removes compute_order_2 from the instruction pipeline to limit the impact of branch misprediction.

Reply:
and in practice we won't have a silent order reduction since we never pass in order > max order ...

So I really included that for third party users or if we make the order an input parameter.

I'm that case, the error mode would be "oops I compiled with too low a shadow factor" and I think it's more efficient to check that outside of the PIC loop.